### PR TITLE
Security config

### PR DESCRIPTION
**Pull request by Ketch92**

### CODE REVIEW

File: src/main/java/core/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    
+    protected void configure(HttpSecurity security) throws Exception {
+        security.authorizeRequests()
+                .anyRequest().authenticated()
+                .and()
+                .formLogin().permitAll()
+                .and()
+                .httpBasic().and()
+                .csrf().disable();
+    }
```
Comment (mfedenko09) [inline src/main/java/core/config/SecurityConfig.java]:
```suggestion
    protected void configure(HttpSecurity security) throws Exception {
        security.authorizeRequests()
        .antMatchers("/register")
                .permitAll()
                .anyRequest().authenticated()
                .and()
                .formLogin().permitAll()
                .and()
                .httpBasic().and()
                .csrf().disable();
    }
```
we can add this if we want enter into /register endpoint without credentials

File: src/main/java/core/controller/ExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException e,
+                                                                  HttpHeaders headers,
+                                                                  HttpStatus status,
+                                                                  WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now().toString());
+        body.put("status", status.value());
+        
+        //TODO check for additional functionality
```
Comment (mfedenko09) [inline src/main/java/core/controller/ExceptionHandler.java]:
```suggestion
```

File: src/main/java/core/controller/OrderController.java [hunk 3]
```diff
@@ -31,15 +33,19 @@ public OrderController(OrderService orderService,
     }
     
     @GetMapping("/orders")
-    public List<OrderResponseDto> getUserOrders(@RequestParam Long userId) {
-        return orderService.getOrdersHistory(userService.get(userId)).stream()
+    public List<OrderResponseDto> getUserOrders(Authentication authentication) {
+        UserDetails details = (UserDetails) authentication.getPrincipal();
+        User user = userService.findByEmail(details.getUsername()).get();
+        return orderService.getOrdersHistory(userService.get(user.getId())).stream()
```
Comment (mfedenko09) [inline src/main/java/core/controller/OrderController.java]:
```suggestion
        return orderService.getOrdersHistory(userService.get(user.getId())).stream()
```
use method findByEmail()

File: src/main/java/core/lib/PasswordValidator.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package core.lib;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import org.springframework.beans.BeanWrapperImpl;
+
+public class PasswordValidator implements ConstraintValidator<FieldMatch, Object> {
```
Comment (mfedenko09) [inline src/main/java/core/lib/PasswordValidator.java]:
```suggestion
public class PasswordValidator implements ConstraintValidator<FieldMatch, UserRequestDto > {
```

File: src/main/java/core/lib/PasswordValidator.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    public void initialize(FieldMatch constraintAnnotation) {
+        this.field = constraintAnnotation.field();
+        this.fieldMatch = constraintAnnotation.fieldMatch();
+    }
+    
+    @Override
+    public boolean isValid(Object obj, ConstraintValidatorContext cvc) {
+        Object fieldValue = new BeanWrapperImpl(obj)
+                .getPropertyValue(field);
+        Object fieldMatchValue = new BeanWrapperImpl(obj)
```
Comment (mfedenko09) [inline src/main/java/core/lib/PasswordValidator.java]:
```suggestion
    public boolean isValid(UserRequestDto userRequestDto, ConstraintValidatorContext cvc) {
        Object fieldValue = new BeanWrapperImpl(userRequestDto)
                .getPropertyValue(field);
        Object fieldMatchValue = new BeanWrapperImpl(userRequestDto)
```

File: src/main/java/core/lib/FieldMatch.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+public @interface FieldMatch {
+    String message() default "Passwords don't match!";
+    
+    String field();
+    
+    String fieldMatch();
+    
+    Class<?>[] groups() default {};
+    
+    Class<? extends Payload>[] payload() default {};
```
Comment (mfedenko09) [inline src/main/java/core/lib/FieldMatch.java:22]:
```suggestion
    Class<? extends Payload>[] payload() default {};
    
    @Target({ElementType.TYPE})
    @Retention(RetentionPolicy.RUNTIME)
    @interface List {
        FieldMatch[] value();
    }
```
Did this work without this code?

File: src/main/java/core/model/dto/UserRequestDto.java [hunk 7]
```diff
@@ -1,8 +1,23 @@
...
+import javax.validation.constraints.Size;
+
+@FieldMatch(message = "Passwords are different.",
+        field = "password",
+        fieldMatch = "repeatPassword")
 public class UserRequestDto {
+    @ValidateEmail
+    @Size(min = 4)
     private String email;
+    @NotNull(message = "Please, don't leave this field empty.")
```
Comment (mfedenko09) [inline src/main/java/core/model/dto/UserRequestDto.java:15]:
```suggestion
```
we check this field on null value in validator class

File: src/main/java/core/model/dto/UserRequestDto.java [hunk 8]
```diff
@@ -1,8 +1,23 @@
...
+    @Size(min = 4)
     private String password;
+    @NotNull(message = "Please, don't leave this field empty.")
```
Comment (mfedenko09) [inline src/main/java/core/model/dto/UserRequestDto.java:18]:
```suggestion
```

**Review by mfedenko09 — CHANGES_REQUESTED** [review]

File: src/main/java/core/lib/PasswordValidator.java (hunk 4 above)
Comment (Ketch92) [reply src/main/java/core/lib/PasswordValidator.java]: I don't think it's a good idea. Furthermore, we can use the same annotation for any field, but if you put there UserRequestDto then we wouldn't be able to do that anymore.

File: src/main/java/core/controller/OrderController.java (hunk 3 above)
Comment (Ketch92) [reply src/main/java/core/controller/OrderController.java]: I use it in the line above. But Forgotten to put it on line 39)  Thanks!

File: src/main/java/core/lib/PasswordValidator.java (hunk 5 above)
Comment (Ketch92) [reply src/main/java/core/lib/PasswordValidator.java]: If I don't change the generic type. There is no need to make these changes

File: src/main/java/core/lib/FieldMatch.java (hunk 6 above)
Comment (Ketch92) [reply src/main/java/core/lib/FieldMatch.java:22]: It works. This part is needed if there are several fields that need to be validated (like 2 passwords, 2 answers for secret questions and so on)

File: src/main/java/core/model/dto/UserRequestDto.java (hunk 7 above)
Comment (Ketch92) [reply src/main/java/core/model/dto/UserRequestDto.java:15]: IDK on your code in the validator class, but in my case there are 2 cases following scenario if password != null we compare password and repeated password in other case (this means password == null) we check if repeated password is also null (and if it is it means password == repeated password == null)

**Review by mfedenko09 — APPROVED** [review]

File: src/main/java/core/config/SecurityConfig.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@Configuration
```
Comment (KseniiaMakarova) [inline src/main/java/core/config/SecurityConfig.java]: we may skip this annotation
@EnableWebSecurity is also a @Configuration

File: src/main/java/core/config/SecurityConfig.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+@EnableWebSecurity
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+    @Autowired
+    public void globalConfig(AuthenticationManagerBuilder managerBuilder) throws Exception {
+        managerBuilder.inMemoryAuthentication()
```
Comment (KseniiaMakarova) [inline src/main/java/core/config/SecurityConfig.java:16]: add passwordEncoder()

File: src/main/java/core/controller/MovieSessionController.java [hunk 11]
```diff
@@ -52,12 +53,13 @@ public MovieSessionResponseDto get(@PathVariable Long id) {
     }
     
     @PostMapping
-    public void addMovieSession(@RequestBody MovieSessionRequestDto requestDto) {
+    public void addMovieSession(@RequestBody @Valid MovieSessionRequestDto requestDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/controller/MovieSessionController.java:56]: what about CinemaHallController?

File: src/main/java/core/lib/EmailValidator.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidator implements ConstraintValidator<ValidateEmail, String> {
+    private static final String PASSWORD_REGEX = "[\\w+@[a-z]+.[a-z]+]";
+    
+    @Override
+    public boolean isValid(String str, ConstraintValidatorContext cvx) {
+        return str != null && !str.matches(PASSWORD_REGEX);
```
Comment (KseniiaMakarova) [inline src/main/java/core/lib/EmailValidator.java]:
```suggestion
        return str != null && str.matches(PASSWORD_REGEX);
```

File: src/main/java/core/lib/FieldMatchValidator.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    private String fieldMatch;
+    
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        this.field = constraintAnnotation.field();
+        this.fieldMatch = constraintAnnotation.fieldMatch();
+    }
+    
+    @Override
+    public boolean isValid(Object obj, ConstraintValidatorContext cvc) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/lib/FieldMatchValidator.java]: let's make parameters informative

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/lib/EmailValidator.java (hunk 12 above)
Comment (Ketch92) [reply src/main/java/core/lib/EmailValidator.java]: Ok.. I had a wrong regex. Thx

File: src/main/java/core/controller/MovieSessionController.java (hunk 11 above)
Comment (Ketch92) [reply src/main/java/core/controller/MovieSessionController.java:56]: Added there as well

File: src/main/java/core/lib/FieldMatchValidator.java (hunk 13 above)
Comment (Ketch92) [reply src/main/java/core/lib/FieldMatchValidator.java]: Ok

File: src/main/java/core/config/SecurityConfig.java (hunk 10 above)
Comment (Ketch92) [reply src/main/java/core/config/SecurityConfig.java:16]: Added

File: src/main/java/core/config/SecurityConfig.java (hunk 9 above)
Comment (Ketch92) [reply src/main/java/core/config/SecurityConfig.java]: I had a trouble there, and tried to solve it with the annotation.

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**